## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [9]:
arr = np.random.randint(0, 10, size=10)
print(arr)
masked = (arr > 3) & (arr < 8)
arr[masked] *= -1
print(arr)


[8 8 7 3 6 3 1 9 3 6]
[ 8  8 -7  3 -6  3  1  9  3 -6]


**2.** Заменить максимальный элемент случайного массива на 0

In [15]:
arr = np.random.randint(0, 20, size=10)
print(arr)
max_index = arr.argmax()
arr[max_index] = 0
print(arr)

[ 4  4 15 12  3 12 18 17 13  9]
[ 4  4 15 12  3 12  0 17 13  9]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [41]:
from itertools import product

arr = np.array([[1, 2], [3, 4]])

# Cartesian product
direct_product = np.array(list(product(*arr)))
print(direct_product)

[[1 3]
 [1 4]
 [2 3]
 [2 4]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [50]:
A = np.random.randint(1, 10, size=(8, 3))
B = np.random.randint(1, 10, size=(2, 2))
print(A)
print(B)
matches = [A[np.all(np.isin(A, b_row), axis=1)] for b_row in B]
for i, m in enumerate(matches):
    print(f"B row {B[i]} matches:\n{m}\n")

[[6 4 2]
 [9 8 1]
 [7 7 6]
 [9 4 5]
 [7 6 1]
 [4 1 9]
 [7 8 7]
 [3 2 6]]
[[9 9]
 [4 2]]
B row [9 9] matches:
[]

B row [4 2] matches:
[]



**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [73]:
C = np.random.randint(1, 3, size=(10, 3))
print(C)
filtered = C[~np.all(C == C[:, 0][:, None], axis=1)]
print('')
print(filtered)


[[1 1 1]
 [2 1 1]
 [1 1 2]
 [1 1 2]
 [2 2 2]
 [2 1 2]
 [1 1 1]
 [2 2 1]
 [2 1 1]
 [2 2 2]]

[[2 1 1]
 [1 1 2]
 [1 1 2]
 [2 1 2]
 [2 2 1]
 [2 1 1]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [76]:
arr = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [1, 2, 3],
    [7, 8, 9],
    [4, 5, 6],
    [10,11,12]
])

unique_rows = np.unique(arr, axis=0)

print("Unique rows:\n", unique_rows)

Unique rows:
 [[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [79]:
X = [[1, 0, 1],
     [2, 0, 2],
     [3, 0, 3],
     [4, 4, 4]]

prod = 1
for i in range(min(len(X), len(X[0]))):
    if X[i][i] != 0:
        prod *= X[i][i]

print(prod) 

3


In [87]:
X = np.array(X)
prod_np = np.prod(X.diagonal()[X.diagonal() != 0])
print(prod_np)

3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [88]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

same_multiset = sorted(x) == sorted(y)
print(same_multiset)

True


In [89]:
from collections import Counter

same_multiset = Counter(x) == Counter(y)
print(same_multiset)

True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [90]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

candidates = x[1:][x[:-1] == 0]

max_preceded_by_zero = np.max(candidates)

print(max_preceded_by_zero) 

5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [91]:
# Input
x = [2, 2, 2, 3, 3, 3, 5]

values = []
counts = []

current = x[0]
count = 1

for i in x[1:]:
    if i == current:
        count += 1
    else:
        values.append(current)
        counts.append(count)
        current = i
        count = 1

# Append the last run
values.append(current)
counts.append(count)

print("Values:", values)
print("Counts:", counts)

Values: [2, 3, 5]
Counts: [3, 3, 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [93]:
import numpy as np
# Sample data
X = np.random.rand(5, 2)
Y = np.random.rand(4, 2)

# Vectorized Euclidean distances
D = np.sqrt(((X[:, np.newaxis, :] - Y[np.newaxis, :, :])**2).sum(axis=2))

print("Distance matrix:\n", D)

Distance matrix:
 [[0.18778586 0.65835932 0.24258422 0.37243018]
 [0.6521747  0.32043483 0.38370967 0.27641477]
 [0.60539978 0.74599929 0.3754284  0.29179941]
 [1.14609495 0.65090294 0.88462404 0.76442994]
 [0.76469663 0.75388116 0.49190374 0.35423427]]


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [1]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [2]:
calorie_stats = np.loadtxt("E:/innowiseHW/homework/pandas_numpy_tasks/data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [17]:
average_competitor = calorie_stats.mean()
average_calories = average_competitor - 60
print(average_calories)

46.883116883116884


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [19]:
calories_stats_sorted = np.sort(calorie_stats)
print(calories_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [21]:
median_calories = np.median(calorie_stats)
median_calories

np.float64(110.0)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [39]:
percentiles = np.percentile(calorie_stats, np.arange(0, 101, 1))
nth_percentile = next(p for p in percentiles if p > 60)
nth_percentile

np.float64(70.0)

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [33]:
more_calories = (calorie_stats > 60).mean() * 100
print("Percentage of cereals >60 calories:", more_calories)

Percentage of cereals >60 calories: 96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [37]:
calories_std = np.std(calorie_stats)
calories_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Analysis of competition cereal shows that average calorie count per serving is around 111, with a median of 110, meaning most products exceed 100 calories. Approximately 96% of competitors have more than 60 calories per serving, nth percentile corresponds to 70